main.go

```diff
@@ -616,6 +308,499 @@ func main() {
+	retryCount := 0
+
+	retryCounterMetric, _ := registerTenantRetryMetric(tCfg)
+
+	for {
+		// Calculate wait time before rerying a failed tenant registration.
+		waitTime := expBackOffWaitTime(retryCount)
+		time.Sleep(time.Duration(waitTime) * time.Millisecond)
+
+		rateLimits := tenantRateLimits(tCfg, t)
+
+		oidcConf, mTLSConf, err := tennatAuthN(t)
+		if err != nil {
+			level.Error(tCfg.logger).Log("msg", "tenant must specify either an OIDC or an mTLS configuration", t.Name, err)
+			retryCounterMetric.With(prometheus.Labels{"tenant": t.Name}).Inc()
+			retryCount++
+
+			continue
+		}
+
+		var (
+			authN       authentication.Middleware
+			authZ       rbac.Authorizer
+			oidcHandler http.Handler
+		)
+
+		if len(oidcConf.Tenant) > 0 {
+			// Create OIDC middleware for a tenant.
+			oidcHandler, authN, err = authentication.NewOIDC(tCfg.logger, "/oidc/"+t.Name, *oidcConf)
+			if err != nil {
+				level.Error(tCfg.logger).Log("msg", "tenant failed to register. retrying ..", t.Name, err)
+				retryCounterMetric.With(prometheus.Labels{"tenant": t.Name}).Inc()
+				retryCount++
+
+				continue
+			}
+		}
+
+		if oidcHandler != nil {
+			r.Mount("/oidc/"+t.Name, oidcHandler)
+		}
+
+		if len(mTLSConf.Tenant) > 0 {
+			mTLSMiddleware := authentication.NewMTLS(*mTLSConf)
+			authN = mTLSMiddleware
+		}
+
+		if t.OPA != nil {
+			authZ = t.OPA.authorizer
+		} else {
+			authZ = tCfg.authorizer
+		}
+
+		if authZ == nil {
+			level.Error(tCfg.logger).Log("msg", "invalid authorization for tenant", t.Name)
+			retryCounterMetric.With(prometheus.Labels{"tenant": t.Name}).Inc()
+			retryCount++
+
+			continue
+		}
+
+		r.HandleFunc("/{tenant}", func(w http.ResponseWriter, r *http.Request) {
+			tenant, ok := authentication.GetTenant(r.Context())
+			if !ok {
+				w.WriteHeader(http.StatusNotFound)
+
+				return
+			}
+			http.Redirect(w, r, path.Join("/api/metrics/v1/", tenant, "graph"), http.StatusMovedPermanently)
+		})
+
+		// Metrics
+		r.Group(func(r chi.Router) {
+			r.Use(authentication.WithTenant(t.Name, t.ID))
+			r.Use(authN)
+			r.Use(authentication.WithTenantHeader(cfg.metrics.tenantHeader, t.ID))
+
+			if tCfg.rateLimitClient != nil {
+				r.Use(ratelimit.WithSharedRateLimiter(tCfg.logger, tCfg.rateLimitClient, rateLimits...))
+			} else {
+				r.Use(ratelimit.WithLocalRateLimiter(rateLimits...))
+			}
+
+			r.Mount("/api/v1/"+t.Name, metricsLegacyHandler(cfg, tCfg, authZ))
+			r.Mount("/api/metrics/v1/"+t.Name, stripTenantPrefix("/api/metrics/v1", metricsHandler(cfg, tCfg, authZ)))
+		})
+		// Logs
+		if cfg.logs.enabled {
+			r.Group(func(r chi.Router) {
+				r.Use(authentication.WithTenant(t.Name, t.ID))
+				r.Use(authN)
+				r.Use(authentication.WithTenantHeader(cfg.logs.tenantHeader, t.ID))
+				r.Mount("/api/logs/v1/"+t.Name, stripTenantPrefix("/api/logs/v1", logsHandler(cfg, tCfg, authZ)))
+			})
+		}
+
+		level.Info(tCfg.logger).Log("msg", "tenant registration is successful :", t.Name)
+
```

Comment:
some counter increment? (: Let's add more observability here since it's now concurrent.